Enabling TLS1.1/1.2 on Android 4.4 devices (API 19/KitKat)

mqus opened the pull request:
This is an attempt to fix #2777 . I already got the Downloader and the Extractor parts working. What I may need some help on is the player. I simply don't know how to approach this and where to start.

Please give me some feedback on the "finished" parts, too.

- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

theScrabi commented:
Player may be complicated as it is a Wohle framework.

mqus commented:
I found https://github.com/google/ExoPlayer/issues/1819 which says I should set `HttpsURLConnection.setDefaultSSLSocketFactory()` "prior to using exoplayer" or implementing a new `DataSource`. Could you comment on what you would find more feasible/desirable? 

Where would be the best place to call that static method and where should I call it if I also want to use it for the other two connections?

As I said, I'm clueless, but motivated. I can do some research myself but I wanted to try the easy way (asking), first ;-).

mqus commented:
I tried the static-function approach and put `HttpsURLConnection.setDefaultSSLSocketFactory()` at the start of MainActivity. I tried it and everything works now. It should not impact other devices which don't run Android 4.4.2 because everywhere I set the new SslFactory I checked for the android version first, but I'm not completely sure. 

In any case, please check the given commits. Should I roll them up into one?

mqus commented:
I'm sorry for my impatience, but Is there a way to speed this process up?

TobiGr commented on app/src/main/java/org/schabi/newpipe/Downloader.java:
```diff
@@ -52,12 +65,41 @@
     }
 
+
+    // Enable TLS 1.2 and 1.1 for OkHTTP on Android Kitkat. This function is mostly taken from the documentation of OkHttpClient.Builder.sslSocketFactory(_,_)
```
Please turn this to a valid Java doc comment by using `/**`.
And please stick to the java code styling conventions and add whitespaces before `{` and after `if` etc. This increases readability significantly and speeds up reviews :)

mqus replied:
Sure! is there a full styleguide available or profiles for android studio/linters? (I'm not that much into android dev)

mqus replied:
found it, i was just being stupid

TobiGr commented on app/src/main/java/org/schabi/newpipe/Downloader.java:
```diff
@@ -52,12 +65,41 @@
+
+            SSLContext sslContext = SSLContext.getInstance("TLS");
+            sslContext.init(null, new TrustManager[] { trustManager }, null);
+            //SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();
```
Please remove this line

TobiGr commented on app/src/main/java/org/schabi/newpipe/util/TLSSocketFactoryCompat.java:
```diff
@@ -0,0 +1,124 @@
+import javax.net.ssl.TrustManager;
+
+
+//This is an extension of the SSLSocketFactory which enables TLS 1.2 and 1.1. Created for usage on Android 4.1-4.4 devices, which don't have enabled those by default.
```
Make this a Java doc comment please :)

TobiGr commented on app/src/main/java/org/schabi/newpipe/util/TLSSocketFactoryCompat.java:
```diff
@@ -0,0 +1,124 @@
+
+    private Socket enableTLSOnSocket(Socket socket) {
+        if(socket != null && (socket instanceof SSLSocket)) {
+            /*
```
Please remove this unused code

mqus replied:
I'm not sure on this one, the way it currently works is to *only* enable 1.1 and 1.2 because soundcloud, youtube and mediaCCC support those (and older ones are frankly insecure). The commented-out code *adds* 1.1 and 1.2 to the existing version array. I'll clean this up as soon as I know what to do.

Redirion replied:
I would prefer the "only enable 1.1 and 1.2" variant

TobiGr replied:
I agree with @Redirion.

mqus replied:
👍

mqus commented:
Thanks for reviewing so far!

I rebased and added most of your suggestions. It should look much cleaner now.

mqus commented:
I now rebased on the current `dev`
I didn't mean to close the PR, force-pushed the wrong stuff accidently.

mqus commented:
So I rebased and tried out PeerTube.

It didn't work. Here is why:
TLS operates with a variable set of cipher suites which have to be matched up when starting a TLS connection to choose one which both client and server support.


The following are enabled by default on an Android 4.4.2 Emulator (All Lists are sorted alphabetically to make comparison easier.)
```
SSL_DHE_DSS_EXPORT_WITH_DES40_CBC_SHA
SSL_DHE_DSS_WITH_DES_CBC_SHA
SSL_DHE_DSS_WITH_3DES_EDE_CBC_SHA
SSL_DHE_RSA_EXPORT_WITH_DES40_CBC_SHA
SSL_DHE_RSA_WITH_DES_CBC_SHA
SSL_DHE_RSA_WITH_3DES_EDE_CBC_SHA
SSL_RSA_EXPORT_WITH_DES40_CBC_SHA
SSL_RSA_EXPORT_WITH_RC4_40_MD5
SSL_RSA_WITH_3DES_EDE_CBC_SHA
SSL_RSA_WITH_DES_CBC_SHA
SSL_RSA_WITH_RC4_128_MD5
SSL_RSA_WITH_RC4_128_SHA
TLS_DHE_DSS_WITH_AES_128_CBC_SHA
TLS_DHE_DSS_WITH_AES_256_CBC_SHA
TLS_DHE_RSA_WITH_AES_128_CBC_SHA
TLS_DHE_RSA_WITH_AES_256_CBC_SHA
TLS_ECDH_ECDSA_WITH_3DES_EDE_CBC_SHA
TLS_ECDH_ECDSA_WITH_AES_128_CBC_SHA
TLS_ECDH_ECDSA_WITH_AES_256_CBC_SHA
TLS_ECDH_ECDSA_WITH_RC4_128_SHA
TLS_ECDH_RSA_WITH_3DES_EDE_CBC_SHA
TLS_ECDH_RSA_WITH_AES_128_CBC_SHA
TLS_ECDH_RSA_WITH_AES_256_CBC_SHA
TLS_ECDH_RSA_WITH_RC4_128_SHA
TLS_ECDHE_ECDSA_WITH_3DES_EDE_CBC_SHA
TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA
TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA
TLS_ECDHE_ECDSA_WITH_RC4_128_SHA
TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA
TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
TLS_ECDHE_RSA_WITH_RC4_128_SHA
TLS_RSA_WITH_AES_128_CBC_SHA
TLS_RSA_WITH_AES_256_CBC_SHA
```
The following are supported on an Android 4.4.2 Emulator, but not enabled:
```
SSL_DH_anon_EXPORT_WITH_DES40_CBC_SHA
SSL_DH_anon_EXPORT_WITH_RC4_40_MD5
SSL_DH_anon_WITH_3DES_EDE_CBC_SHA
SSL_DH_anon_WITH_DES_CBC_SHA
SSL_DH_anon_WITH_RC4_128_MD5
SSL_RSA_WITH_NULL_MD5
SSL_RSA_WITH_NULL_SHA
TLS_DH_anon_WITH_AES_128_CBC_SHA
TLS_DH_anon_WITH_AES_256_CBC_SHA
TLS_ECDH_anon_WITH_RC4_128_SHA
TLS_ECDH_anon_WITH_AES_128_CBC_SHA
TLS_ECDH_anon_WITH_AES_256_CBC_SHA
TLS_ECDH_anon_WITH_3DES_EDE_CBC_SHA
TLS_ECDH_anon_WITH_NULL_SHA
TLS_ECDH_ECDSA_WITH_NULL_SHA
TLS_ECDH_RSA_WITH_NULL_SHA
TLS_ECDHE_ECDSA_WITH_NULL_SHA
TLS_ECDHE_RSA_WITH_NULL_SHA
```
(these are either `anon`, which means they are unsuitable for TLS or `NULL` which means unencrypted, so none are desirable for us)

Those will be enabled by the last commit, if possible: (OkHttp moderntls + `TLS_ECDHE_ECDSA_WITH_AES_{128,256}_CBC_SHA)`, adapted from https://github.com/AntennaPod/AntennaPod/pull/3439/files#diff-f0eab34b36f9ab46c533bdd783fddc46:
```
SSL_RSA_WITH_3DES_EDE_CBC_SHA
TLS_AES_128_GCM_SHA256
TLS_AES_256_GCM_SHA384
TLS_AES_128_CCM_SHA256
TLS_AES_256_CCM_8_SHA256
TLS_CHACHA20_POLY1305_SHA256
TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA
TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA
TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256
TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256
TLS_RSA_WITH_AES_128_GCM_SHA256
TLS_RSA_WITH_AES_256_GCM_SHA384
TLS_RSA_WITH_AES_128_CBC_SHA
TLS_RSA_WITH_AES_256_CBC_SHA
```

On an Android 4.4.2 device without vendor changes to the suites, this means that the following cipher suites are enabled with tls1.2:
```
TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA
TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA
TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
TLS_RSA_WITH_AES_128_CBC_SHA
TLS_RSA_WITH_AES_256_CBC_SHA
```

Which doesn't overlap with the [Framatube.org cipher suites](https://www.ssllabs.com/ssltest/analyze.html?d=framatube.org&s=178.63.240.148#suitesBody), so no connection can be made. If you notice, it does say that kitkat devices can connect below but I simply can't find the source for that statement; I also couldn't enable the given Ciphersuite in any way.

If we want to fix that on the server (we probably have to), we could start by changing the default nginx config recommended by Peertube and changed here: (https://github.com/Chocobozzz/PeerTube/commit/e883399fa6caa56bb8519c9a2e22d88001f26661)

mqus commented:
There was another thing I noticed, while trying out mediaCCC some more:

When I download a video, it will work at first but then stall, while the console throws the following error(with context):
```
D/DownloadMission: 0:[request]  Range=bytes=22544384-23068671
    0:[response] Code=206
    0:[response] Content-Length=524288
    0:[response] Content-Range=bytes 22544384-23068671/61431254
D/DownloadMission: 2:[request]  Range=bytes=21495808-22020095
    2:[response] Code=206
    2:[response] Content-Length=524288
    2:[response] Content-Range=bytes 21495808-22020095/61431254
D/DownloadRunnable: 0:position 43 stopped 23068672/23068671
D/dalvikvm: GC_FOR_ALLOC freed 6166K, 35% free 34766K/53176K, paused 40ms, total 40ms
D/DownloadMission: 1:[request]  Range=bytes=22020096-22544383
    1:[response] Code=206
    1:[response] Content-Length=524288
    1:[response] Content-Range=bytes 22020096-22544383/61431254
D/DownloadRunnable: 0:acquired block at position=44 done=0
D/DownloadMission: 0:[request]  Range=bytes=23068672-23592959
    0:[response] Code=206
    0:[response] Content-Length=524288
    0:[response] Content-Range=bytes 23068672-23592959/61431254
D/DownloadRunnable: 0:position 44 stopped 23592960/23592959
    0:acquired block at position=45 done=0
D/DownloadRunnable: 2:position 41 stopped 22020096/22020095
    2:acquired block at position=46 done=0
E/DownloadMission: notifyError()
    javax.net.ssl.SSLHandshakeException: javax.net.ssl.SSLProtocolException: SSL handshake aborted: ssl=0x65da54e8: Failure in SSL library, usually a protocol error
    error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure (external/openssl/ssl/s23_clnt.c:741 0x5fb14d74:0x00000000)
        at com.android.org.conscrypt.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:448)
        at com.android.okhttp.Connection.upgradeToTls(Connection.java:146)
        at com.android.okhttp.Connection.connect(Connection.java:107)
        at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:294)
        at com.android.okhttp.internal.http.HttpEngine.sendSocketRequest(HttpEngine.java:255)
        at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:206)
        at com.android.okhttp.internal.http.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:345)
        at com.android.okhttp.internal.http.HttpURLConnectionImpl.getResponse(HttpURLConnectionImpl.java:296)
        at com.android.okhttp.internal.http.HttpURLConnectionImpl.getResponseCode(HttpURLConnectionImpl.java:503)
        at com.android.okhttp.internal.http.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:136)
        at us.shandian.giga.get.DownloadMission.establishConnection(DownloadMission.java:249)
        at us.shandian.giga.get.DownloadRunnable.run(DownloadRunnable.java:84)
     Caused by: javax.net.ssl.SSLProtocolException: SSL handshake aborted: ssl=0x65da54e8: Failure in SSL library, usually a protocol error
    error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure (external/openssl/ssl/s23_clnt.c:741 0x5fb14d74:0x00000000)
        at com.android.org.conscrypt.NativeCrypto.SSL_do_handshake(Native Method)
        at com.android.org.conscrypt.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:405)
        	... 11 more
    notifyError() code = 1004
I/DownloadManagerService: Active network [connected=true metered=false] NetworkInfo: type: WIFI[], state: CONNECTED/CONNECTED, reason: (unspecified), extra: "TomateMozarella v4", roaming: false, failover: false, isAvailable: true, smCause: 0, isConnectedToProvisioningNetwork: false
W/DownloadMission: thread alive: DownloadMission[1] Immer noch Nichts zu verbergen_.mp4
D/DownloadMission: 2:[request]  Range=bytes=24117248-24641535
    2:[response] Code=206
    2:[response] Content-Length=524288
    2:[response] Content-Range=bytes 24117248-24641535/61431254
D/DownloadRunnable: thread 2 exited from main download loop
    The mission has been paused. Passing.
W/DownloadMission: thread alive: DownloadMission[2] Immer noch Nichts zu verbergen_.mp4
D/MissionAdapter: checkMasterButtonsVisibility() running=true paused=false
I/ViewRootImpl: ViewRoot's Touch Event : Touch Down
D/DownloadRunnable: 0:acquired block at position=42 done=0
D/DownloadRunnable: 2:acquired block at position=45 done=0
D/DownloadRunnable: 1:acquired block at position=46 done=0
D/MissionAdapter: checkMasterButtonsVisibility() running=true paused=false
I/ViewRootImpl: ViewRoot's Touch Event : Touch UP
D/DownloadMission: 2:[request]  Range=bytes=23592960-24117247
    2:[response] Code=206
    2:[response] Content-Length=524288
    2:[response] Content-Range=bytes 23592960-24117247/61431254
D/DownloadMission: 1:[request]  Range=bytes=24117248-24641535
    1:[response] Code=206
    1:[response] Content-Length=524288
    1:[response] Content-Range=bytes 24117248-24641535/61431254
D/DownloadMission: 0:[request]  Range=bytes=22020096-22544383
    0:[response] Code=206
    0:[response] Content-Length=524288
    0:[response] Content-Range=bytes 22020096-22544383/61431254
V/AudioSystem: ioConfigChanged() event 0, ioHandle 3
    ioConfigChanged() opening already existing output! 3
V/AudioSystem: ioConfigChanged() event 0, ioHandle 2
    ioConfigChanged() opening already existing output! 2
I/ViewRootImpl: ViewRoot's Touch Event : Touch Down
I/ViewRootImpl: ViewRoot's Touch Event : Touch UP
I/ActivityManager: Timeline: Activity_launch_request id:org.schabi.newpipe.debug time:141260754
```

I can pause and resume the download and it will continue where it stopped, so It's not a huge issue but it 'll be annoying. I'll investigate that some more.

mqus commented:
I tried watching with wireshark and only found one mirror (selfhost.de) having an unspecified TLS error while connecting... I don't know how to fix this so i'll let it be for now.

yausername commented:
@mqus [this](https://github.com/AntennaPod/AntennaPod/issues/2814#issuecomment-426367766) comment says that changing the variable name of `SSLSocketFactory` in `TLSSocketFactoryCompat` to `delegate` might help. No idea why though.

yausername commented:
After a reread I understand that peertube won't work on kitkat unless the instances update their `ssl_ciphers`.

mqus commented:
I tried your `delegate` suggestion (I read about it myself and then later forgot ;) ) and it doesn't seem to do anything for the second problem either. I guess that it worked in a previous version of okhttp but isn't neccessary now that you have to supply a TrustManager when setting a SSLSocketFactory... . I think the second problem arises from Kitkat having a pretty old root cert store while a few sites depend on a new one... but I'm not entirely sure.

mqus commented:
Interestingly, the peertube docker config contains a traefik instance, which would work with kitkat: https://github.com/Chocobozzz/PeerTube/blob/develop/support/docker/production/config/traefik.toml#L27 
It enables `TLS_RSA_WITH_AES_256_CBC_SHA`(0x35) which ssllabs considers weak, but not insecure. It's also in the above list of Ciphers which okhttp will try.

mqus commented:
For the meantime, I tried another instance (tube.tchncs.de) and tested downloading and playing. 

While browsing works, downloading obviously only works for videos from instances with similar TLS settings. 
But: playing videos doesn't work at all, it throws a wall of the following repeating logstream with errors (https://pastebin.com/c901HP13) while constantly showing the toaster message "Recovering from player error".  This is similar when viewing videos of other instance. I think this isn't a case of TLS issues (?)

For instances with "wrong" tls settings,  that doesn't happen and it will instead repeat the message `W/SimpleExoPlayer: Player is accessed on the wrong thread. See https://exoplayer.dev/issues/player-accessed-on-wrong-thread` slowly. Downloading from those instances will repeatedly throw SSLHandShakeErrors (or something similar) as expected.

mqus commented:
With help from @Chocobozzz, framatube.org is working flawlessly on Android 4.4 and the upstream nginx template was altered (https://github.com/Chocobozzz/PeerTube/pull/2330)!

I checked video playback again and found that this is mostly an emulator issue. On the x86 Kitkat emulator without play services, video playback was not possible. On my personal phone it works just fine though.

As this thread has gotten a bit long and unwieldy, I'll summarize the past and present issues with newpipe on Kitkat:

### MediaCCC:
- [x] Browse: fixed
- [x] Playback: fixed
  - Doesn't work in my Kitkat emulator(x86, no play services), but on my device.
- [x] Download: mostly fixed
  - still minor issues with some mirrors, I don't know how to fix them or if they appear on other Android versions, works otherwise

### Peertube instances *with* matching ciphersuites (set server-side):
- [x] Browse: fixed
- [x] Playback: fixed
  - Doesn't work in my Kitkat emulator(x86, no play services), but on my device. 
- [x] Download: fixed
- This now includes framatube.org (Thanks, @framasoft !)

### Peertube instances *without* matching ciphersuites (set server-side):
- [ ] Browse: doesn't work
  - works via other instances, but it's not easy to figure out from which instance a video/channel is hosted
- [ ] Playback
- [ ] Download

This is now in a state I would consider finished. The last big group has to be solved by writing kind emails to the admins of affected instances and maybe tune the Peertube UI a bit. If there are no other opinions for the commented-out code in the reviewed section, I'll clean it up and this is ready to get merged.

yausername commented:
Great work @mqus!
@TobiGr Can we include this in the latest release? I believe you would have to merge this into `release_0.18.0` branch.

TobiGr commented on app/src/main/java/org/schabi/newpipe/DownloaderImpl.java:
```diff
@@ -153,4 +168,34 @@ public Response execute(@Nonnull Request request) throws IOException, ReCaptchaE
+
+            SSLContext sslContext = SSLContext.getInstance("TLS");
+            sslContext.init(null, new TrustManager[] { trustManager }, null);
+            //SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();
```
unused code

TobiGr commented on app/src/main/java/org/schabi/newpipe/DownloaderImpl.java:
```diff
@@ -153,4 +170,47 @@ public Response execute(@Nonnull Request request) throws IOException, ReCaptchaE
+
+            builder.connectionSpecs(Arrays.asList(legacyTLS, ConnectionSpec.CLEARTEXT));
+        } catch (KeyManagementException | NoSuchAlgorithmException | KeyStoreException e) {
+            e.printStackTrace();
```
Please only print to the console, when debug mode is enabled. You can _import_ the `DEBUG` constant from https://github.com/TeamNewPipe/NewPipe/blob/88a6e5981f6fc7cc5c9caae7a0d0b492efa4ee89/app/src/main/java/org/schabi/newpipe/MainActivity.java#L82

TobiGr commented on app/src/main/java/org/schabi/newpipe/util/TLSSocketFactoryCompat.java:
```diff
@@ -0,0 +1,128 @@
+        try {
+            HttpsURLConnection.setDefaultSSLSocketFactory(getInstance());
+        } catch (NoSuchAlgorithmException | KeyManagementException e) {
+            e.printStackTrace();
```
same here. In case you want to open an ErrorScreen by using  one of the `reportError()` methods from  https://github.com/TeamNewPipe/NewPipe/blob/dev/app/src/main/java/org/schabi/newpipe/report/ErrorActivity.java

mqus commented:
Rebase&fixing is done. Github still wants me to adress your concerns but wont let me resolve them...

mqus commented:
Thank you @ all! If there are any connection errors with kitkat devices in the future or any other stuff I should reproduce with my phone, just ping me and I'll see what I can do. I'll also write some more E-Mails to instance admins but I probably wont cover them all.

TobiGr commented:
@mqus thank you for the effort! I'll gladly come back to your offer :heart:
